### PR TITLE
Get rid of weird hack, comint-delchar-or-maybe-eof

### LINKED ISSUE #914: C-d in shell-mode + evil-normal-state should do evil-scroll-down
As title. comint-mode has something called `comint-delchar-or-maybe-eof`, which is bound to `C-d` in shell-mode (and probably other modes based on comint-mode).

The idea is that `C-d` usually signifies the EOF character when typing in a shell, so comint detects whether point is at the input prompt in the buffer. If so, then `C-d` should send an EOF character.

Unfortunately, it is implemented in such a way that if you are not at the input prompt, then it is hard-coded to fall back to `delete-char`, which is the function that is bound to `C-d` by default in Emacs. Instead, it should fall back to whatever `C-d` was bound to by the user's settings. In this case, that would mean `evil-scroll-down`.

It seems like this is probably a comint-mode bug rather than a spacemacs bug, but it would be nice to at least have a workaround in spacemacs of some kind since comint-mode is part of the main Emacs distribution and probably won't change fast.

### PR DESCRIPTION
**Pull request by kini**
This fixes #914.

### CODE REVIEW

File: spacemacs/config.el [hunk 1]
```diff
@@ -168,6 +168,12 @@ Can be installed with `brew install trash'."
 ;; Single space between sentencs is more widespread than double
 (setq-default sentence-end-double-space nil)
 
+;; The C-d rebinding that most shell-like buffers inherit from
+;; comint-mode assumes non-evil configuration with its
+;; `comint-delchar-or-maybe-eof' function, so we disable it
+(with-eval-after-load 'comint
```
Comment (syl20bnr) [inline spacemacs/config.el]: `with-eval-after-load` is not supported in 24.3 so we have to use `eval-after-load`

**Comment by syl20bnr** [comment]
What weird hack this PR gets rid of ? ?
Is there any weird side effect like `C-d` not working in insert state ?

**Comment by kini** [comment]
I was just referring to `comint-delchar-or-maybe-eof` as a weird hack. `coment-delchar-or-maybe-eof` is bound to `C-d` in comint-mode. It checks whether point is at the prompt, and if so, it sends an EOF character. If point is not at the prompt, then it runs `delete-char`. The reason it runs `delete-char` is that `delete-char` is the usual binding of `C-d` in Emacs, but if we're using evil-mode, that is not the case. `comint-delchar-or-maybe-eof` was supposed to only change the behavior of `C-d` if point is at the prompt, but it ends up changing the behavior of `C-d` in other places in the buffer as well, which is why I called it a weird hack.

> Is there any weird side effect like `C-d` not working in insert state ?

`C-d` will not send an EOF character anymore in insert state. It will run whatever is bound to `C-d` in insert state, which by default is `evil-shift-left-line`. But if you want to send `C-d` then you can type `C-c C-d` just as you could before.

File: spacemacs/config.el (hunk 1 above)
Comment (kini) [inline spacemacs/config.el]: I see. Then, why didn't Travis catch it? Should we change the `.travis.yml` to test against Emacs 24.3?

**Comment by syl20bnr** [comment]
Thank you for the explanation, I did not know for `C-c C-d`:-D

Thank you for the fix as well :+1: 
Cherry-picked into develop branch, you can safely delete your branch.

**Pull request closed without merging**